Add zones to Hostgroups and Servicegroups

eurotux-tec opened the pull request:
In the scenario of our company there is an icinga master node and several satellites, each one is responsible for monitoring the infrastructure of a specific customer.

Currently, all hostgroups / servicegroups are replicated to all satellites, so all customers are aware of the other hostgroups / servigroups, which should be avoided.

This PR contains changes to the PHP code and DB schema to allow a zone to be assigned to a hostgroup / servicegroup, so that the hostgroup / servicegroup configuration file is only sent to the corresponding zone.

Do you think that this makes sense or can we do it in a different way? If you agree, please merge this. If you don't, please let me know how can I improve the patch so that is gets merged.

araujorm commented:
Hello.

Sorry for having had this stalled for so long, but I made the corrections that were asked, leaving only what is *indeed* relevant, adapted to be on par with the current master version (uuid property added, database version and so on), and refreshed the branch used for this pull request.

Can you review it, and if good, ponder about merging it? Anything that might be missing please tell.

Thanks for your work.

araujorm commented:
> Any objection against unifying Service/Host/UserGroup code for Objects and Forms, at least as far as possible? I kind of dislike the duplicated code. Parts of it already used to be there, but now it get's even more.

@Thomas-Gelf Had missed this part. Meanwhile, took a look a the code and tried to unify it.

We added a new commit to the pull request (in order to have it clear what was made to add support for zones on hostgroups and service groups, and what was made to unify the code afterwards). 

For objects, moved some common parts to IcingaObjectGroup, where zone_id is now in defaultProperties and thus those don't need to be overriden in IcingaHostGroup and IcingaServiceGroup anymore. Similar for relations property, which don't need to be specified in descendant classes anymore.

For forms, created a intermediate class IcingaGroupForm (analog to the already existing IcingaObjectGroup for objects) which extends DirectorObjectForm, and made IcingaHostGroupForm,  IcingaServiceGroupForm and IcingaUserGroupForm extend this new class instead. Moved the addZoneElements method to this new class (although it's not used by IcingaUserGroupForm), thus removing it from the descendant classes that had it repeated.

Was this what you had in mind, @Thomas-Gelf ? If you think this should be on a separate pull request, we can remove this new commit and make another pull request later after the zone support for hostgroups and servicegroups is merged. Please let us know, will await your feedback.

araujorm commented:
Made a correction on the last commit message, because the addZoneElements method **is** used by all the descendant classes, including IcingaUserGroupForm.

Hope this is what is intended, best regards.

araujorm committed:
Add zones to Hostgroups and Servicegroups

Correspondingly update DB schema and add migrations

araujorm committed:
Unify Service/Host/UserGroup code for Objects and Forms

At least as far as possible.

For objects, moved some common parts to IcingaObjectGroup, where zone_id
is now in defaultProperties and thus those don't need to be overriden in
IcingaHostGroup and IcingaServiceGroup anymore. Similar for relations
property, which don't need to be specified in descendant classes
anymore.

For forms, created a intermediate class IcingaGroupForm (analog to the
already existing IcingaObjectGroup for objects) which extends
DirectorObjectForm, and made IcingaHostGroupForm,
IcingaServiceGroupForm and IcingaUserGroupForm extend this new class
instead. Moved the addZoneElements method to this new class (although
it's not used by IcingaUserGroupForm), thus removing it from the
descendant classes that had it repeated.

eurotux-tec commented:
Hello.

We've refreshed the proposed commits so that they merge with the current master.

Could we have any feedback on whether it's acceptable, or if there is any other approach you'd prefer instead?

araujorm commented:
Good afternoon. Branch of the merge request was rebased so that the MR is valid for latest release and master branch.
